### DISCUSSION TITLE
Derived state as functions?

### DISCUSSION THREAD

**Discussion by rvision** [pmndrs/zustand#1809] [Q&A] [answered]
Hi, I'm fairly new to zustand, used redux/redux toolkit before and I am testing zustand with Immer for the new project. 

I've read elsewhere that derived state can be problematic. I've tried to put functions that return derived state and this works fine.

Is there something I'm doing wrong here? Is this approach valid? Thanks for any suggestion.

the store (simple wizard steps):

```
const useConfigWizardStore = create(
	immer((set, get) => ({
		currentStep: 1,
		totalSteps,
		completedSteps: new Set(),
		isFirstStep: () => get().currentStep === 1,
		isLastStep: () => get().currentStep === totalSteps,
		isFinished: (id = null) => {
			if (id === null) {
				return get().completedSteps.size === totalSteps;
			}
			return get().completedSteps.has(id);
		},
		actions: {
			completeStep: (id) =>
				set((state) => {
					state.completedSteps.add(id);
				}),
			nextStep: () =>
				set((state) => {
					if (state.currentStep === totalSteps) {
						return;
					}
					state.currentStep = state.currentStep + 1;
				}),
			prevStep: () =>
				set((state) => {
					if (state.currentStep === 1) {
						return;
					}
					state.currentStep = state.currentStep - 1;
				})
		}
	}))
);
```

actual use in components:
```
const isFinished = useConfigWizardStore((state) => state.isFinished());
const isFirstStep = useConfigWizardStore((state) => state.isFirstStep());
const isLastStep = useConfigWizardStore((state) => state.isLastStep());
const isFinishedStep3 = useConfigWizardStore((state) => state.isFinished(3));
```

**Comment by dai-shi** [comment, marked as answer]
This is actually not a derived state as we discussed somewhere else. It's essentially a selector function.

As best practice, we recommend typical selector usage.
```js
const isFinished = useConfigWizardStore((state) => state.currentStep === 1);
```

But you can use your solution if that feels more comfortable. (I just don't know possible pitfalls. The code looks identical, so it should be fine.)

**Reply by rvision** [reply]
Daishi, thanks a lot for fast response!

You're right, this is a selector, not derived state.

It's more comfortable because I would like to encapsulate logic related to the state operations. Also, reusability - isFinished will be copied around or should go into separate (another hook) or it can be as selector, closer to the state, which I prefer.

I don't know what the fuss is all about derived state then? I mean - you can create a custom hook that consumes 2 state slices and wrap it in useMemo for the end consuming component. I don't see any problems with this approach?

Thanks again, zustand is awesome!

**Reply by dai-shi** [reply]
AFAICT, when people say derived states, they assume the computation is cached. It's not about problem. It's about optimization. Here's my experimental lib: https://github.com/dai-shi/derive-zustand

**Reply by rvision** [reply]
Well, that's why I mentioned useMemo - computation will be cached. Or I am missing something?

**Reply by dai-shi** [reply]
useMemo caches only in a component. It can't cache things across multiple components.

**Reply by dai-shi** [reply]
btw, if someone needs such memoization, I'd suggest to look into [proxy-memoize](https://github.com/dai-shi/proxy-memoize).

**Reply by rvision** [reply]
> useMemo caches only in a component. It can't cache things across multiple components.

Hm, which brings us back to prop-drilling. 

If you have one, let's say, root component and pass the cached useMemo state, it will work. Otherwise - not.
